### ISSUE TITLE
Wrong keyof behaviour for generic with extends types in 2.9

### ISSUE THREAD

**Issue by mctep** [microsoft/TypeScript#24560]
**TypeScript Version:**  2.9.1, 3.0.0-dev.20180601

**Search Terms:** keyof 2.9 generics extract extends string number symbol

**Code**

```ts
type StringKeyof<T> = Extract<keyof T, string>;

type Omit<T, K extends StringKeyof<T>> = any;

type WithoutFoo = Omit<{ foo: string }, "foo">; // ok

type WithoutFooGeneric<P extends { foo: string }> = Omit<P, "foo">; // Error: Type '"foo"' does not satisfy the constraint 'Extract<keyof P, string>'.
```

**Expected behavior:**

`Omit` should pass `"foo"` as valid type for keyof because generic type extends a type that has this string key.

**Actual behavior:**

Error: `Type '"foo"' does not satisfy the constraint 'Extract<keyof P, string>'.`

**Playground Link:** [Link](https://www.typescriptlang.org/play/index.html#src=type%20StringKeyof%3CT%3E%20%3D%20Extract%3Ckeyof%20T%2C%20string%3E%3B%0D%0A%0D%0Atype%20Omit%3CT%2C%20K%20extends%20StringKeyof%3CT%3E%3E%20%3D%20any%3B%0D%0A%0D%0Atype%20WithoutFoo%20%3D%20Omit%3C%7B%20foo%3A%20string%20%7D%2C%20%22foo%22%3E%3B%0D%0A%0D%0Atype%20WithoutFooGeneric%3CP%20extends%20%7B%20foo%3A%20string%20%7D%3E%20%3D%20Omit%3CP%2C%20%22foo%22%3E%3B%0D%0A%0D%0A)

**Comment by 7sempra** [comment]
This is also true for function-bound generics:

    interface Foo {
        foo: number,
    }

    function run<T extends Foo>() {
        // ERROR: `Type '"foo"' is not assignable to type Extract<keyof T, string>`
        let key: Extract<keyof T, string> = 'foo';

        // Okay
        let key2: Extract<keyof Foo, string> = 'foo';
    }

[Playground](http://www.typescriptlang.org/play/#src=%0D%0Ainterface%20Foo%20%7B%0D%0A%20%20%20%20foo%3A%20number%2C%0D%0A%7D%0D%0A%0D%0Afunction%20doWork%3CT%20extends%20Foo%3E()%20%7B%0D%0A%20%20%20%20%2F%2F%20ERROR%3A%20%60Type%20'%22foo%22'%20is%20not%20assignable%20to%20type%20Extract%3Ckeyof%20T%2C%20string%3E%60%0D%0A%20%20%20%20let%20key%3A%20Extract%3Ckeyof%20T%2C%20string%3E%20%3D%20'foo'%3B%0D%0A%0D%0A%20%20%20%20%2F%2F%20Okay%0D%0A%20%20%20%20let%20key2%3A%20Extract%3Ckeyof%20Foo%2C%20string%3E%20%3D%20'foo'%3B%0D%0A%7D%0D%0A)

**Comment by weswigham** [comment]
`let key: Extract<keyof T, string> = 'foo';` shouldn't work. Substitute `never` for `T` (which is allowable), and you'd be attempting to assign `'foo'` to `never`, which shouldn't work.

**Comment by 7sempra** [comment]
Huh, interesting, how is that different from using a bare `keyof`?

    function run<T extends Foo>() {
        // Okay
        let key: keyof T = 'foo';
    }

If T is `never` in this case, `keyof T` is also `never`, so we're still assigning `'foo'` to `never`.

**Comment by hpohlmeyer** [comment]
Is this the same issue?

```tsx
type FormFields = { someField: string };

// Typescript should allow only keys of "FormFields" to be
// used as fieldName 
class GenericForm<T extends FormFields> {
    public fieldA = new FormField<T>({ fieldNameA: "someField" }); // works as expected
    public fieldB =  new FormField<T>({ fieldNameB: "someField" }); // throws an error
}

interface FormFieldOptions<T extends FormFields> {
    fieldNameA?: keyof T;
    fieldNameB?: Extract<keyof T, string>;
}
class FormField<T extends FormFields> {
    public fieldNameA: string;
    public fieldNameB: string;

    constructor(options: FormFieldOptions<T>) {
        // Does not work because fieldNameA is of type "symbol | number | string"
        this.fieldNameA = options.fieldNameA;

        // Does work because fieldNameB is ensured to be a string 
        this.fieldNameB = options.fieldNameB;
    }
}
```

[Playground Link](https://www.typescriptlang.org/play/#src=type%20FormFields%20%3D%20%7B%20someField%3A%20string%20%7D%3B%0A%0A%2F%2F%20Typescript%20should%20allow%20only%20keys%20of%20%22FormFields%22%20to%20be%0A%2F%2F%20used%20as%20fieldName%20%0Aclass%20GenericForm%3CT%20extends%20FormFields%3E%20%7B%0A%20%20%20%20public%20fieldA%20%3D%20new%20FormField%3CT%3E(%7B%20fieldNameA%3A%20%22someField%22%20%7D)%3B%20%2F%2F%20works%20as%20expected%0A%20%20%20%20public%20fieldB%20%3D%20%20new%20FormField%3CT%3E(%7B%20fieldNameB%3A%20%22someField%22%20%7D)%3B%20%2F%2F%20throws%20an%20error%0A%7D%0A%0Ainterface%20FormFieldOptions%3CT%20extends%20FormFields%3E%20%7B%0A%20%20%20%20fieldNameA%3F%3A%20keyof%20T%3B%0A%20%20%20%20fieldNameB%3F%3A%20Extract%3Ckeyof%20T%2C%20string%3E%3B%0A%7D%0Aclass%20FormField%3CT%20extends%20FormFields%3E%20%7B%0A%20%20%20%20public%20fieldNameA%3A%20string%3B%0A%20%20%20%20public%20fieldNameB%3A%20string%3B%0A%0A%20%20%20%20constructor(options%3A%20FormFieldOptions%3CT%3E)%20%7B%0A%20%20%20%20%20%20%20%20%2F%2F%20Does%20not%20work%20because%20fieldNameA%20is%20of%20type%20%22symbol%20%7C%20number%20%7C%20string%22%0A%20%20%20%20%20%20%20%20this.fieldNameA%20%3D%20options.fieldNameA%3B%0A%0A%20%20%20%20%20%20%20%20%2F%2F%20Does%20work%20because%20fieldNameB%20is%20ensured%20to%20be%20a%20string%20%0A%20%20%20%20%20%20%20%20this.fieldNameB%20%3D%20options.fieldNameB%3B%0A%20%20%20%20%7D%0A%7D)

**Comment by weswigham** [comment]
Please note: my older comment on this issue is incorrect - `keyof never` is now `string | number | symbol`, so `Extract`ing anything from `unknown` is just the thing extracted.

**Comment by amoscatelli** [comment]
Is this the same issue ?

`function run<T extends { foo: string }>() {
  let key: {
    [K in keyof { foo: string }]:
    { foo: string }[K] extends String ? K : never
  }[keyof { foo: string }];
  key = 'foo'; // OK

  let key2: {
    [K in keyof T]:
    T[K] extends String ? K : never
  }[keyof T];
  key2 = 'foo'; // NOT OK
}`

**Comment by jack-williams** [comment]
@amoscatelli No, the issue there is that conditional types ignore the constraints on type parameters when checking assignability, so `T[K] extends string ? K : never` does not get simplified to `K`.

**Comment by princefishthrower** [comment]
Knock knock three years later 😄 

What is the update, if any, on this? `Extract` with `keyof Something` works as expected for any concrete `Something`, but as soon as you throw generics into the mix `Extract` doesn't enforce anything at all... it still is effectively `keyof T`. I feel like IntelliSense should be able to figure this out, i.e. looking at a simple generic `isAGreaterThanB` function:

```ts
// only extract those types from keyof T which make sense with the '<' operator
function isAGreaterThanB <T>(a: T, b: T, compareOnProperty: Extract<keyof T, string | Date | number>) {
  return a[compareOnProperty] > b[compareOnProperty];
}
```

[Full Playground example.](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgJIGUD2BbCYAWoA5sgN4CwAUMjcjJpgFzIDOYUxA3FbcgEZwozACJxI3arQEAvZiACu2PtAm9sATwBi8kAjDBMIZgAoAlMgC8APmQA3TMAAmq2hoDyfAFYQ9zCpN4aeiZWdi4eWgBfFxoNdHUlTAAbZhYEvmSJKkiqKgRDNlYcPEIQIgBBZgxigmJLMgighmYAInwIJKTMZAB3TCgkxxaAGkb+QTkIHuRRSDNRgPHZZAAGBbUtHT0DI2QzSxt8kBZkiAA6LqJjAHJ2zu6+gcdr03XXdQ9vXwbF3mDmW4dLq9fqDa5jSJvWLqeKJFLIWEZJJmbK5ShHQonXC1MoAISqWGxpRIFh+f2ayDaQIeoKGUPGQmQICmMzEEHmYxkzAAjPSNNpdPpDCZzNZkBjThdMFdAfcQU8XnyPl4fGA-GNySFZcDHmCIUrEclmIbkaZUZQqAB6S3IQxJdTICAAD3YiDAyAImBYKDA6gADhAWHQoDhkABrCDqTAwZAAFV6hAQ+GQ2DgEdYEGOKB6wAIHvayGuAB5rraA1AxP0qDAtkKQMhgCxygBxKAQNlQWP4OAgXHIIuxqzGODMWPDfij8f5bB+wQQNwgAAKIfLvuYAFEXRW9EWI1GY2PQhwysgAD6syBnpmKZRQKzmfw0NtgeRQetwADa09nbYXy8wq7qAAusgNh8F+OA-vOS4rtAvpARIOQWpQ1rIOU9w9BAjiMHkBSSpcxiNi2bYdl2Pa4sYWIlMQ5TjlROJELi44tMELSmKYEgSkk5wEURrbtpAnbdr2lE1MStFFESxBMZSAhQGxHG4cc+HSoRTb8aRwkUfR4l0WJ0nMTIClZMhqHoPgmDyIMIDXO6yjIHAGFYThKE2sY6GevIRDJmI+aNsgfoOOAcb+hA6AIBwfruiwFlWY4GY+t27rxv51RSWUACEyAAOpJQ2QZEEFJCGAA-BlZquXs6EnA5IDqD0cAOvAwBJEGvlQDo+i4HQ-T5ig1xxOkySlggcDehVXE8apfEkYJZEiTpNF6eljHMfytY7MZSknNxUpXDNAnQPN2n6WUEmLXia3Kl8YBbeieG7bx6mzUdWmiSt52natlKDXCxlAA)

**Comment by carlvincentld** [comment]
@princefishthrower Your issue is not related to the original issue. I think you might have misunderstood the usage of the keyword [`Extract`](https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union) or [`keyof T`](https://www.typescriptlang.org/docs/handbook/2/keyof-types.html#the-keyof-type-operator).

Taking back the interface `ISomething` with slight alteration to the names:
```typescript
interface ISomething {
    foo: string;
    bar: Date;
    baz: number;
    qux: () => void;
    quux: { foo: string; };
    corge: symbol;
}

type Keys = keyof ISomething; // 'foo' | 'bar' | 'baz' | 'qux' | 'quux' | 'corge';
type ExtractOfKeys = Extract<Keys, string | Date | number>; // 'foo' | 'bar' | 'baz' | 'qux' | 'quux' | 'corge';
```

Since the keys of `ISomething` are all strings, `Extract` returns the full list of keys.

My guess is you want only the keys foo, bar or baz, since they are of ordered types. Instead of `Extract`, you might be looking to use the following "FieldOfType" type:
```typescript
type FieldOfType<Type, ValueType> = { [K in keyof Type]: Type[K] extends ValueType ? K : never }[keyof Type];
function isAGgreaterThanB<T>(a: T, b: T, compareOnProperty: FieldOfType<T, string | Date | number>) {
    return a[compareOnProperty] > b[compareOnProperty];
}

type FieldsOfType = FieldOfType<ISomething, string | Date | number>; // 'foo' | 'bar' | 'baz'
```

Here's a [Playground Link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgJIGUD2BbCYAWoA5sgN4CwAUMjcjJpgFzIDOYUxA3FbcgEZwozACJxI3arQEAvZiACu2PtAm8AjvIAezABQBKZAF4AfMgBumYABNVtDVuak6DZmw4ginZAF9bNBJhQRBCuAJ5KmAA2Et5UVFYQCJGCKAEgbKw4eIQeAILMGFkEXPGJyVCpmOlgmbjFHgBCBVh1OZ5xlGChAA4oAGLAEJFWAPIwACo9EAA8k70ANMgAanCR8hBzEKaGZMgA2gDSyKDIANYQoZgwyJsAusybh7fIEJqQIFYsy6vrm8gA-MgjnIIGZoD49udLtc7hIqAB6eHIADqwAIyGCIGgwAQyC6vWQOlAMGgX0KrWIeioMHkIAQYGAVQxFTE0HG+DgIFmxh0cAeiz4-OQ3SgmF6UC6zAGQ1GEyms0WbmIyAAPshRJBVcgFEpoMY9Mw+AxIhBOWQeLQKmB5FAQMg4HsRWLoF1nqY+I7ReLXTEqGkMvRMEZmabIFB2ZydCwim1coqY8QGosAOSB5N6CT+moCKDBogssMRkBRhN5eMUxopnPpzNVDIyPMFtkc4vRitEOO1bKJqtwaQ1hFIgBymEgzGTGk0yZT9in9o+yGTASCEGT9oqwq90EioWQCUg9IgVntX1AZlW1j9dZqk8boebkbb3bLXfqRCTi8nA8oWeQs7vrLhi2Jbtp2T5vh+E7yFo36-suwQAYWwHgbG5bPu+KbwauGYdLhlB4YiKJovgyBpAgVooPiKA6GRFFoC06FUpQNJ0gyTL5veQGcgxb68s0pZEAK-HtosTreqEUqDMMYybNM5LoYq7DKmqGooGqOrKFA+qGsapp2hQkg0FaNp2g6YkuqEbr8J6zoSpZvo-teziYAATIhD4gDxbQgehYECZBaY4Y51T8IIbk7BxgFFl5xA+W+fntpB1ZBb+MjhSGUUtjFHhxahr5tElfawU5k7pZFSHcQJuXEAl6GQV+KUldBmhlU2XGeVVKE1WhEEzs1xUhVhrWcdFnUCbVvWLlh354bNlCEQAqiwyqvOwiA1DRVTkXgKDyW+TEsfSjJ2uVHkAKJvFA618fRAlCbdImkTg3QpCMIAAApbnZzAXWt9LTFCVwPQprBKR4WqqVqGl6ga-C6WaBm8IR0CikINz4CgtliIEyCmAgnIgKO-AoHA3TdJEgzHmAQZUYuLDhEakTJhaRl4CZ9p7AE2AvRUb2fbZrq49ZXM8xAfNfT6VCxMFAYMAAzO57W-Vd9LVS+XWVougW1iFOYKxFbVFsr10ax2PUFb2UADfWfb6xlFUgMbqumxNFuLjI36ESOY6fjBfUwfOx5LoEwRriwEAQNgXzU8TJ7LUQIBwHwJp4jTGN0LSR1VIOvC0AA7oQkQZAABqd7XF3uO30tHGPYCexwgOeFNWAAdFeIWTnbZdG5dJvjebPa+1OjUd81XeGy2TtgGrZv5YPUEwSPGRYePI2T73zv93PmvByuM34QfBFIkaCgsJJMoyVMyB54EpxfHnxGsAzURnBcp7XAQFzINgpOp8gVSYw-unBIy0KhUymCwduGR6Z1x2OgZ+kR9A62geEdK8CIiIKCqAMM8AkBoFRAQdBjNzSGX2DA+42pFCaT8GQ1BFClQeBoT-MmxBxiYCIVEMIGCHIJCSCkJ6IUYH5HwcRDh0RSh8I3L+GBTQRGEIQXCSg3dkLhDAuED8MCgrKMfKoxU6i9HYBcloieOjsBqOwJBZh3RWHsIQTWIAA) if you want to play with it.

**Comment by RyanCavanaugh** [comment]
We're very unequipped to deal with this sort of code.

The core question being posed this code is, in essence, *does any possible instantiation of `WithoutFooGeneric` produce a type such that `Extract<keyof P, string>` produces a supertype of `"foo"` ?*

That's sort of an easy question when the constraint is a simple object type like `{ "foo": string }`. But I don't see a tractable path forward in the general case. Consider other constraints you could write:
```ts
type WithoutFooGeneric<P extends Partial<Record<"foo" | "bar", string>>> = Omit2<P, "foo">;
```

This should be an error, since writing
```ts
type A = WithoutFooGeneric<Record<"bar", string>>
```
violates the constraint of `Omit2`

But how is this error *even found*? It's not through instantiation of the constraint; that won't work - that fails to catch the error here
```ts
// k: "foo" | "bar"
type K = Partial<Record<"foo" | "bar", string>>;
```

Given an arbitrary constraint `T extends C`, there are very few things that can be readily determined about any possible instantiation of `T` -- you can't just use `C` as-is and get correct answers out of the other side, as shown by the `Partial` example. We have some logic in place when relating `T` as a subtype, but once you take a contravariant operation like `keyof T`, all bets are off.

So TL;DR this isn't the kind of thing someone can just put up a bugfix for - it implies a very large new kind of reasoning to be able to fix beyond the specific case presented here (aside: fixing only specific cases usually makes everything *much worse* later).
